packages/loopring_v3/contracts/amm/libamm/AmmBlockReceiver.sol

```diff
@@ -0,0 +1,127 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2017 Loopring Technology Limited.
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "../../core/impl/libtransactions/BlockReader.sol";
+import "./AmmData.sol";
+import "./AmmExitProcess.sol";
+import "./AmmJoinProcess.sol";
+import "./AmmUpdateProcess.sol";
+
+
+/// @title AmmBlockReceiver
+library AmmBlockReceiver
+{
+    using AmmExitProcess    for AmmData.State;
+    using AmmJoinProcess    for AmmData.State;
+    using AmmUpdateProcess  for AmmData.State;
+    using BlockReader       for ExchangeData.Block;
+
+    function beforeBlockSubmitted(
+        AmmData.State      storage S,
+        uint                       poolTokenTotalSupply,
+        ExchangeData.Block memory  _block,
+        uint                       txIdx,
+        bytes              memory  auxiliaryData
+        )
+        public
+        returns (uint)
+    {
+        AmmData.PoolTransaction[] memory poolTransactions = abi.decode(
+            auxiliaryData,
+            (AmmData.PoolTransaction[])
+        );
+
+        // Cache the domain seperator to save on SLOADs each time it is accessed.
+        uint size = S.tokens.length;
+        AmmData.Context memory ctx = AmmData.Context({
+            _block: _block,
+            txIdx: txIdx,
+            domainSeperator: S.domainSeperator,
+            exchangeDomainSeparator: S.exchange.getDomainSeparator(),
+            ammActualL2Balances: new uint96[](size),
+            ammExpectedL2Balances: new uint96[](size),
+            numTransactionsConsumed: 0,
+            tokens: S.tokens,
+            size: size,
+            poolTokenTotalSupply: poolTokenTotalSupply,
```

Comment:
While we can cache this, it needs to be updated for each mint/burn as they happen to do all calculations correctly. In theory the total supply state variable only needs to be read/updated once instead of for each mint/burn, but that makes things a bit more complicated (multiples updates to the same storage aren't cheap but also not super expensive, it's 1,600 gas/update for the SLOAD + SSTORE).

Reply:
remove this cache.